internal/attestation/azure/snp/validator.go

```diff
@@ -36,249 +41,359 @@ type Validator struct {
+	vcek, vcekParsingErr := a.parseVCEK()
 	if err != nil {
-		return fmt.Errorf("marshalling ecdsa signature: %w", err)
+		logger.Warnf("Error parsing VCEK: %v", vcekParsingErr)
+	}
+	if vcek != nil {
+		att.CertificateChain.VcekCert = vcek.Raw
+	}
+	// Otherwise, retrieve it from AMD KDS.
+	if att.CertificateChain.VcekCert == nil {
+		logger.Infof("VCEK certificate not present, falling back to retrieving it from AMD KDS")
+		vcekURL := kds.VCEKCertURL(productName, report.GetChipId(), kds.TCBVersion(report.GetReportedTcb()))
+		vcek, err := getter.Get(vcekURL)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving VCEK certificate from AMD KDS: %w", err)
+		}
+		att.CertificateChain.VcekCert = vcek
+	}
+
+	// If the certificate chain is present, parse it and format it.
+	ask, ark, certChainParsingErr := a.parseCertChain()
+	if certChainParsingErr != nil {
+		logger.Warnf("Error parsing certificate chain: %v", certChainParsingErr)
+	}
+	if ask != nil {
+		att.CertificateChain.AskCert = ask.Raw
+	}
+	if ark != nil {
+		att.CertificateChain.ArkCert = ark.Raw
+	}
+	// Otherwise, retrieve it from AMD KDS.
+	if att.CertificateChain.AskCert == nil || att.CertificateChain.ArkCert == nil {
+		logger.Infof(
+			"Certificate chain not fully present (ARK present: %t, ASK present: %t), falling back to retrieving it from AMD KDS",
+			(att.CertificateChain.ArkCert != nil),
+			(att.CertificateChain.AskCert != nil),
+		)
+		kdsCertChain, err := trust.GetProductChain(productName, getter)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving certificate chain from AMD KDS: %w", err)
+		}
+		if att.CertificateChain.AskCert == nil {
+			att.CertificateChain.AskCert = kdsCertChain.Ask.Raw
+		}
+		if att.CertificateChain.ArkCert == nil {
+			att.CertificateChain.ArkCert = kdsCertChain.Ark.Raw
+		}
 	}
 
-	buf := new(bytes.Buffer)
-	if err = binary.Write(buf, binary.LittleEndian, report); err != nil {
-		return fmt.Errorf("writing report to buf: %w", err)
-	}
-	// signature is only calculated from 0x0 to 0x2a0
-	if err := cert.CheckSignature(x509.ECDSAWithSHA384, buf.Bytes()[:0x2a0], sigEncoded); err != nil {
-		return &signatureError{err}
-	}
+	return att, nil
+}
 
-	hasExpectedIDKeyDigest := false
-	for _, digest := range v.config.FirmwareSignerConfig.AcceptedKeyDigests {
-		if bytes.Equal(digest, report.IDKeyDigest[:]) {
-			hasExpectedIDKeyDigest = true
-			break
+// parseCertChain parses the certificate chain from the instanceInfo into x509-formatted ASK and ARK certificates.
+// If less than 2 certificates are present, only the present certificate is returned.
+// If more than 2 certificates are present, an error is returned.
+func (a *azureInstanceInfo) parseCertChain() (ask, ark *x509.Certificate, retErr error) {
+	newlinesTrimmed := strings.TrimSpace(string(a.CertChain))
+
+	i := 1
+	var rest []byte
+	var block *pem.Block
+	for block, rest = pem.Decode([]byte(newlinesTrimmed)); block != nil; block, rest = pem.Decode(rest) {
+		if i > 3 {
```

Comment:
```suggestion
		if i > 2 {
```

Reply:
I think this should be correct here? We expect 2 PEM blocks to be present, and `i` is initialized with `1`. So we would only want to error on the third certificate, i.e. from `i >= 4`

Reply:
1st iteration (1st cert): i=1, i>2==false
2nd iteration (2nd cert): i=2, i>2==false
3rd iteration (3rd cert): i=3, i>2==true